### ISSUE TITLE
Space insert before formatted text not working when not in first line

### ISSUE THREAD

**Issue by mzorz** [wordpress-mobile/AztecEditor-Android#605]
This seems like a regression of #530, addressed by #555. 

### Expected
Space insert should never break formatting

### Observed
Space insertion breaks formatting when it's done on the second or N+1 lines

### Reproduced
1. get on an empty AztecText (i.e. demo app without content)
2. start typing "he"
3. toggle Bold
4. type "llo"
5. place the cursor right before the bold-formatted text, like "he|**llo**" and tap SPACE
6. observe the bold text remains in bold, while the non-formatted text remains unstyled, like "he **llo**" which is expected.
7. place the cursor at the end of the line and hit ENTER
8. on this new line, repeat steps 2 to 5.
9. observe that the second line loses formatting (everything is unstyled, like "he llo").

#### Tested
[Google Pixel] on [8.0, API 26] with [latest develop demo]

**Comment by mzorz** [comment]
Adding some logging, I could see in steps 8 and 9 the TextWatchers are a bit crazy and their callbacks seem to be being executed out of order. Commenting out the addition of `EndOfBufferMarkerAdder.install()` seemed to bring it back to normal and actually makes this bug disappear (that is, the buffered detection mechanism implemented in #584 #595 is able to detect the pattern all times, as the sequence of events comes in as expected), so I'm further studying how `EndOfBufferMarkerAdder` works and will likely try to come up with a solution that includes what it's currently doing, but avoiding producing nested `TextWatcher` calls as happens from making changes to `text`  in the `afterTextChanged` callback (which `EndOfBufferMarkerAdder` does).

**Comment by mzorz** [comment]
> Adding some logging, I could see in steps 8 and 9 the TextWatchers are a bit crazy and their callbacks seem to be being executed out of order

adding the logs for the full sequence of the second situation here for self-evidence tracking (note the new lines here are exactly what the content of the text is, as we're trying to identify what happens when we perform the experiment on a second line):

```
01-15 16:17:18.755 1403-1403/? I/LatinMotionHandler: consumeEvent() : GestureTypingEnabled = true
01-15 16:17:19.542 7652-7652/org.wordpress.aztec D/WordPress-EDITOR: beforeTextChanged: "ho la
                                                                     hola" start: 6 count: 2after: 0
01-15 16:17:19.544 7652-7652/org.wordpress.aztec D/WordPress-EDITOR: onTextChanged: "ho la
                                                                     la" start: 6 count: 0
01-15 16:17:19.554 7652-7652/org.wordpress.aztec D/WordPress-EDITOR: onTextChanged: "ho la
                                                                     la" start: 6 count: 0
01-15 16:17:19.564 7652-7652/org.wordpress.aztec D/WordPress-EDITOR: afterTextChanged: "ho la
                                                                     la"
01-15 16:17:19.578 7652-7652/org.wordpress.aztec D/WordPress-EDITOR: beforeTextChanged: "ho la
                                                                     la" start: 6 count: 2after: 0
01-15 16:17:19.580 7652-7652/org.wordpress.aztec D/WordPress-EDITOR: onTextChanged: "ho la
                                                                     " start: 6 count: 0
01-15 16:17:19.582 7652-7652/org.wordpress.aztec D/WordPress-EDITOR: onTextChanged: "ho la
                                                                     " start: 6 count: 0
01-15 16:17:19.593 7652-7652/org.wordpress.aztec D/WordPress-EDITOR: beforeTextChanged: "ho la
                                                                     " start: 6 count: 0after: 1
01-15 16:17:19.594 7652-7652/org.wordpress.aztec D/WordPress-EDITOR: onTextChanged: "ho la
                                                                     ​" start: 6 count: 1
01-15 16:17:19.597 7652-7652/org.wordpress.aztec D/WordPress-EDITOR: onTextChanged: "ho la
                                                                     ​" start: 6 count: 1
01-15 16:17:19.602 7652-7652/org.wordpress.aztec D/WordPress-EDITOR: afterTextChanged: "ho la
                                                                     ​"
01-15 16:17:19.607 7652-7652/org.wordpress.aztec D/WordPress-EDITOR: afterTextChanged: "ho la
                                                                     ​"
01-15 16:17:19.616 7652-7652/org.wordpress.aztec D/WordPress-EDITOR: beforeTextChanged: "ho la
                                                                     ​" start: 6 count: 0after: 3
01-15 16:17:19.617 7652-7652/org.wordpress.aztec D/WordPress-EDITOR: onTextChanged: "ho la
                                                                     ho ​" start: 6 count: 3
01-15 16:17:19.623 7652-7652/org.wordpress.aztec D/WordPress-EDITOR: onTextChanged: "ho la
                                                                     ho ​" start: 6 count: 3
01-15 16:17:19.634 7652-7652/org.wordpress.aztec D/WordPress-EDITOR: beforeTextChanged: "ho la
                                                                     ho ​" start: 9 count: 1after: 0
01-15 16:17:19.635 7652-7652/org.wordpress.aztec D/WordPress-EDITOR: onTextChanged: "ho la
                                                                     ho " start: 9 count: 0
01-15 16:17:19.636 7652-7652/org.wordpress.aztec D/WordPress-EDITOR: onTextChanged: "ho la
                                                                     ho " start: 9 count: 0
01-15 16:17:19.641 7652-7652/org.wordpress.aztec D/WordPress-EDITOR: afterTextChanged: "ho la
                                                                     ho "
01-15 16:17:19.642 7652-7652/org.wordpress.aztec D/WordPress-EDITOR: afterTextChanged: "ho la
                                                                     ho "
01-15 16:17:19.651 7652-7652/org.wordpress.aztec D/WordPress-EDITOR: beforeTextChanged: "ho la
                                                                     ho " start: 9 count: 0after: 2
01-15 16:17:19.654 7652-7652/org.wordpress.aztec D/WordPress-EDITOR: onTextChanged: "ho la
                                                                     ho la" start: 9 count: 2
01-15 16:17:19.656 7652-7652/org.wordpress.aztec D/WordPress-EDITOR: onTextChanged: "ho la
                                                                     ho la" start: 9 count: 2
01-15 16:17:19.660 7652-7652/org.wordpress.aztec D/WordPress-EDITOR: afterTextChanged: "ho la
                                                                     ho la"

```


see this part specifically, things do not come in the `before`, `on`, and `after` expected order, but rather appear in a different order because of changes being triggered / nested recursively in  `EndOfBufferMarkerAdder`:

```
01-15 16:17:19.578 7652-7652/org.wordpress.aztec D/WordPress-EDITOR: beforeTextChanged: "ho la
                                                                     la" start: 6 count: 2after: 0
01-15 16:17:19.580 7652-7652/org.wordpress.aztec D/WordPress-EDITOR: onTextChanged: "ho la
                                                                     " start: 6 count: 0
01-15 16:17:19.582 7652-7652/org.wordpress.aztec D/WordPress-EDITOR: onTextChanged: "ho la
                                                                     " start: 6 count: 0
01-15 16:17:19.593 7652-7652/org.wordpress.aztec D/WordPress-EDITOR: beforeTextChanged: "ho la
                                                                     " start: 6 count: 0after: 1
01-15 16:17:19.594 7652-7652/org.wordpress.aztec D/WordPress-EDITOR: onTextChanged: "ho la
                                                                     ​" start: 6 count: 1
01-15 16:17:19.597 7652-7652/org.wordpress.aztec D/WordPress-EDITOR: onTextChanged: "ho la
                                                                     ​" start: 6 count: 1
01-15 16:17:19.602 7652-7652/org.wordpress.aztec D/WordPress-EDITOR: afterTextChanged: "ho la
                                                                     ​"
01-15 16:17:19.607 7652-7652/org.wordpress.aztec D/WordPress-EDITOR: afterTextChanged: "ho la
```
